### PR TITLE
Webhooks review draft

### PR DESCRIPTION
**Pull request by keharper**
## Purpose of this pull request

This pull request (PR) adds the following topics to the webhooks doc set

Hook configuration
Response and logging
webhooks.xml reference

Other topics updated as needed.

Staging build: https://developer-stage.adobe.com/commerce/extensibility/webhooks/

## Affected pages

- ...

## Links to Magento Open Source code

- ...

### CODE REVIEW

File: src/pages/webhooks/xml-schema.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+| `resolver` | String  | The resolver class that appends headers dynamically.         | false       | Not applicable     |
+| `remove`   | Boolean | Set to `true` to remove the header from the request.           | false       | false   |
+
+## fields and field elements
+
+A `fields` element is optional and can contain one or more `field` elements. The `fields` element provides the ability to limit the payload of a webhook to only those fields defined in the individual `field` definitions.
+
+| Attribute | Type   | Description  | Is required | Default |
+|---|---|---|---|---|
+| `name` | String  | The path to the field to include, such as `product.sku`. Do not include the top-level `data` object in the name. | true        | Not applicable     |
```
Comment (ryanjcohen) [inline src/pages/webhooks/xml-schema.md:94]: Think my answers on this were confusing. There are cases where `data` should be included in the name. For example, a field like
`<field name='data.product.sku'/>` is valid for an `observer.catalog_product_save_after event `- the unfiltered payload for the event has a top-level `data` key and in this case the name value is being used as the path to the value in the unfiltered payload and as the desired path to the value in the filtered payload.

For plugin events, many wouldn't have payloads with a top-level data key and therefore if the source attribute isn't added it doesn't make much sense to include the top-level data in the name value, but it's not technically disallowed

I'm not sure of the best terms to use, but also wondering if it would be helpful to add to the description that the name is the path in the processed/filtered payload and the source is the path in the unprocessed/unfiltered webhook data payload

File: src/pages/webhooks/commands.md [hunk 2]
```diff
@@ -7,20 +7,41 @@ keywords:
...
+
+The value of the `webhook-name` argument must be a valid Commerce event name. The value must use the following pattern:
+
+```text
+<type>.<event_name>
+```
+
+where:
+
+*  `type` specifies the origin of the event. Specify `observer` if the event is emitted by a Commerce observer, or specify `plugin` if the event is emitted by a plugin.
```
Comment (oshmyheliuk) [inline src/pages/webhooks/commands.md:33]: `or specify `plugin` if the event is emitted by a plugin.`

I would rephrase it to 
`or specify `plugin` if the event is based on methods of Resource Models or Api Interfaces`

As it's not emitted by the plugin, we are generating plugins for Resource Models or API Interfaces based on those event names for.

File: src/pages/webhooks/hooks.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+# Configure hooks
+
+A hook defines a request to a remote server. The `hook` element of the `webhooks.xml` file describes the communication parameters, including the URL of the remote service to execute, HTTP method, and timeout information. Additional configuration occurs within the `headers` and `fields` elements. Each individual `header` element defines an HTTP header key/value pair, while the individual `field` elements define what data is transmitted to the external server.
+
+The [webhook configuration reference](xml-schema.md) describes each element in detail.
+
+## Define the connection
+
+The following hook definition includes the hook name, the destination URL, and timeout information. The `required` attribute indicates a that a failure will terminate the process. The `fallbackErrorMessage` will be written to the error log and can be displayed on the storefront.
```
Comment (oshmyheliuk) [inline src/pages/webhooks/hooks.md:16]: `The `required` attribute indicates a that a failure will terminate the process.`
maybe we can mention that by default all hooks are required. 
And seems `a` is redundant in `indicates a that`

File: src/pages/webhooks/hooks.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+This example defines two headers:
+
+Name | Value
+--- | ---
+`Authorization` | `Bearer:` and the value of the `APP_VALIDATE_STOCK_AUTHORIZATION_TOKEN` environment variable.
+`api-key` | Value from the Adobe Commerce configuration to the file containing the key (`path/to/api-key`).
+
+### Dynamic header resolvers
+
+You might not want to store secrets that expire in environment variables. In these cases, create a dynamic header resolver to manage these values.vTo create your own resolver, define a new class that implements `Magento\AdobeCommerceWebhooks\Model\HeaderResolverInterface`, as shown below.
```
Comment (oshmyheliuk) [inline src/pages/webhooks/hooks.md:49]: `.vTo create your own resolver,` should be a space instead of `v`

File: src/pages/webhooks/hooks.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+```xml
+<fields>
+    <field name="customer.entity_id" source="data.customer.entity_id" />
+    <field name="customer.customer_email" source="context_customer_session.get_customer.get_email" />
+</fields>
+```
+
+In this example, the value of `customer.customer_email` will be set to `Magento\Customer\Model\Session::getCustomer()::getEmail()`. If the value does not exist or cannot be processed, the appropriate message will be logged, and the hook execution will not be interrupted.
+
+You can also specify the string arguments to use when accessing the application context. Provide these arguments within curly braces and delimit them with colons, if there are multiple arguments. The following example sets `config_value` to the value of `Magento\Framework\App\Config\ScopeConfigInterface::getValue('value/path', 'default')`.
```
Comment (jhadobe) [inline src/pages/webhooks/hooks.md:165]: Does `if there are multiple arguments` apply to both the colons and the curly braces? Might want to clarify.

File: src/pages/webhooks/responses.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
+---
+title: Webhooks responses and logging
+description: Learn how to process webhook responses and review logged errors.
+keywords:
+  - Extensibility
+---
+
+# Webhook responses and logging
+
+Currently, Adobe Commerce webhooks support only two responses: `success` and `exception`. Responses that require additional processing, such as parsing the results of a GET response, are not supported.
```
Comment (jhadobe) [inline src/pages/webhooks/responses.md:10]:
```suggestion
Currently, Adobe Commerce webhooks only support two responses: `success` and `exception`. Responses that require additional processing, such as parsing the results of a GET response, are not supported.
```

File: src/pages/webhooks/responses.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+Exceptions and notices are logged in the `<installation_directory>/var/log/system.log` file.
+
+## Responses
+
+The endpoint is expected to return a 200 response and a JSON object that indicates the result of the operation. The object can contain the following fields:
+
+Field | Description
+--- | ---
+`op` | The status of the operation. Must contain either `success` or `exception`.
```
Comment (jhadobe) [inline src/pages/webhooks/responses.md:20]: `contain` seems strange here.

File: src/pages/webhooks/responses.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+`class` | If the `op` status is `exception`, optionally specifies the exception class. If not set, the `\Magento\Framework\Exception\LocalizedException` will be thrown.
```
Comment (jhadobe) [inline src/pages/webhooks/responses.md:21]:
```suggestion
`class` | If the `op` status is `exception`, optionally specifies the exception class. If `class` is not set, the `\Magento\Framework\Exception\LocalizedException` will be thrown.
```

File: src/pages/webhooks/responses.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+`message` |  If the `op` status is `exception`, optionally specifies the exception message. If not set, then the message defined in the `fallbackErrorMessage` configuration field will be returned. If `fallbackErrorMessage` is not set, the system default error message will be shown.
```
Comment (jhadobe) [inline src/pages/webhooks/responses.md:22]:
```suggestion
`message` |  If the `op` status is `exception`, optionally specifies the exception message. If `message` is not set, then the message defined in the `fallbackErrorMessage` configuration field will be returned. If `fallbackErrorMessage` is not set, the system default error message will be shown.
```

File: src/pages/webhooks/responses.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+# Webhook responses and logging
+
+Currently, Adobe Commerce webhooks support only two responses: `success` and `exception`. Responses that require additional processing, such as parsing the results of a GET response, are not supported.
+
+Exceptions and notices are logged in the `<installation_directory>/var/log/system.log` file.
+
+## Responses
+
+The endpoint is expected to return a 200 response and a JSON object that indicates the result of the operation. The object can contain the following fields:
```
Comment (jhadobe) [inline src/pages/webhooks/responses.md:16]:
```suggestion
The endpoint is expected to return a `200` response and a JSON object that indicates the result of the operation. The object can contain the following fields:
```

File: src/pages/webhooks/responses.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+The `exception` operation causes Commerce to terminate the process that triggered the original event. The exception is logged in `<installation_directory>/var/log/system.log`.
+
+## Logging
+
+The following table describes webhook logging activity. Each hook can be configured to have its own soft and hard timeout values.
+
+| Case | Logging activity
+| --- | --- |
```
Comment (jhadobe) [inline src/pages/webhooks/responses.md:51]: It might be easier to visually scan this table if the columns were switched.

File: src/pages/webhooks/xml-schema.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
+---
+title: Webhooks configuration reference
+description: Reference information for constructing a webhooks.xml file
```
Comment (jhadobe) [inline src/pages/webhooks/xml-schema.md:3]: I think we need a sentence description for SEO.

File: src/pages/webhooks/xml-schema.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+keywords:
+  - Extensibility
+---
+
+# Webhook configuration reference
+
+Webhooks are configured in a `webhooks.xml` file. This file can be placed in the system `<install_directory>/app/etc` directory or in the `etc` directory of an enabled Adobe Commerce module.
+
+Run the `bin/magento webhooks:list` command  to determine if the webhooks you create in this file register correctly. The webhook name will be displayed in the list of the registered webhooks.
```
Comment (jhadobe) [inline src/pages/webhooks/xml-schema.md:12]:
```suggestion
Run the `bin/magento webhooks:list` command to determine if the webhooks you create in this file register correctly. The webhook name will be displayed in the list of registered webhooks.
```

File: src/pages/webhooks/xml-schema.md [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        |__ hooks
+            |__ batch
+                |__ hook
+                    |__ headers
+                    |   |__ header
+                    |__ fields
+                        |__ field
+```
+
+## config attributes
```
Comment (jhadobe) [inline src/pages/webhooks/xml-schema.md:28]: Are you not doing `code` tags in headings in this repo?

File: src/pages/webhooks/xml-schema.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+## method element
+
+The `method` element must define the webhook `name` and `type`. The combination of method name and type must be unique across the system.
+
+| Attribute | Type  | Description | Is required | Default |
+|---|---|---|---|
+| `name`    | String | The webhook code name. The value must be in the form `<type>.<webhook_name>`, where `type` is either `observer` or `plugin`, and `webhook_name` matches a valid Commerce event name. Use the `bin/magento webhooks:list:all` command to display a list of possible webhooks. | true        | Not applicable     |
+| `type`    | String | Specifies whether the webhook should be executed `before` or  `after` the original action.  | true        | Not applicable     |
```
Comment (jhadobe) [inline src/pages/webhooks/xml-schema.md:46]:
```suggestion
| `type`    | String | Specifies whether the webhook should be executed `before` or `after` the original action.  | true        | Not applicable     |
```

**Review by jhadobe — CHANGES_REQUESTED** [review]

File: src/pages/webhooks/xml-schema.md (hunk 1 above)
Comment (keharper) [reply src/pages/webhooks/xml-schema.md]: I forgot to update this line after it became clear to me that you have to know the shape of the data you're getting in the default webhook and the shape of the data you're sending. I changed the example in the "Configure hooks" topic so that it changes the shape and renames a field. So, I think my confusion when we were discussing this area has ben cleared up.

**Review by jhadobe — APPROVED** [review]

File: src/pages/webhooks/commands.md [hunk 16]
```diff
@@ -256,3 +278,55 @@ bin/magento webhooks:generate:module
 ```terminal
 Module was generated in the app/code/Magento directory
 ```
+
+## Emulate webhook execution
+
+The `webhooks:dev:run` command is used for development and testing purposes only. It emulates the execution of your registered webhook containing a custom payload with requiring changes to the Commerce application.
+Run this command after setting the initial webhook payload in a `webhooks.xml` file, Then run the command again any time you make subsequent modifications to the payload until you can confirm that the payload works as expected.
```
Comment (mariam21arauj) [inline src/pages/webhooks/commands.md:285]: This is very small and it may be disregarded if not necessary to change, but there's a capital T after a comma: "...payload in a webhooks.xml file, Then...". I think there should either be a period to keep the capital T there, or make it a lowercase t. I think second option makes more sense.

Comment (keharper) [reply src/pages/webhooks/commands.md]: @mariam21arauj Comments/PRs to fix typos are always welcomed. Thanks!

File: src/pages/webhooks/index.md [hunk 17]
```diff
@@ -7,370 +7,30 @@ keywords:
...
-| `remove`    | Boolean | Set to true to remove the field from the payload.                                     | false       | false   |
-
-#### Hook headers attributes:
-
-| Attribute  | Type    | Description                                                  | Is required | Default |
-|------------|---------|--------------------------------------------------------------|-------------|---------|
-| `name`     | String  | The header name will be in the same form as it will be sent. | false       | n/a     |
-| `resolver` | String  | The resolver class that appends headers dynamically.         | false       | n/a     |
-| `remove`   | Boolean | Set to true to remove the header from the request.           | false       | false   |
+1. [Test](testing.md) your `webhooks.xml` file and understand the [response](responses.md).
```
Comment (jhadobe) [inline src/pages/webhooks/index.md:36]: Is "understand" the right word here? maybe they should "validate" the response or something similar.

File: src/pages/webhooks/testing.md [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    </method>
+```
+
+Instead of manually adding a product to the cart from the storefront, you can run the following command, which specifies a custom payload:
+
+```bash
+bin/magento webhooks:dev:run observer.checkout_cart_product_add_before:before '{"data":{"product":{"sku":"simple-product","name":"Simple Product"}}}'
+```
+
+The emulated webhook endpoint receives the following payload, according to `fields` configured for the webhook:
```
Comment (jhadobe) [inline src/pages/webhooks/testing.md:35]:
```suggestion
The emulated webhook endpoint receives the following payload, according to the `fields` configured for the webhook:
```
